### PR TITLE
Fix XQuery 3.1 casting and numeric comparison compliance

### PR DESCRIPTION
**Pull request by joewiz**
## Summary

Push XQuery 3.1 XQTS compliance (HEAD test suite) from **85.3% to 88.8%** (+652 tests) by fixing error codes, URI resolution, and parser edge cases.

### Casting & type fixes
- **Implement XQuery 3.1 casting table** (`Type.isCastable`): Pre-validates cast operations. Impossible casts now correctly raise XPTY0004 instead of FORG0001. Fixes ~580 tests.
- **Fix fn:min/fn:max mixed numeric comparisons**: Allow cross-numeric-family comparisons by checking numeric union membership. Fixes ~60 tests.
- **Fix FOCA0002 for NaN/INF → integer/decimal**: NaN/INF casts now correctly raise FOCA0002 instead of FORG0001. Fixes ~44 tests.
- **Fix xs:boolean → integer subtypes**: Route through IntegerValue for proper range validation. Fixes ~5 tests.
- **Fix generic ERROR codes to FORG0001**: StringValue.checkType() and Base64BinaryValueType now use ErrorCodes.FORG0001. Fixes ~58 tests.

### fn:json-doc URI resolution
- **Fix relative URI resolution**: Resolve relative URIs against the static base URI before falling back to the database URI prefix. Fixes ~295 tests (misc-JsonTestSuite: 0% → 93%).
- **Fix FOJS0001 error code**: JSON parse errors (JsonParseException) now correctly raise FOJS0001 instead of FOUT1170.

### fn:parse-ietf-date improvements
- **Case-insensitive day/month names**: Accept "SAT", "aug", etc.
- **Single-digit hours**: Grammar allows 1-2 digits.
- **Optional whitespace before timezone**: Allow "19:36:01GMT".
- **24:00:00 normalization**: Normalize to 00:00:00 of the next day.
- Fixes ~8 tests.

### XQTS HEAD Results

| Metric | Baseline | After |
|--------|----------|-------|
| Pass rate | 85.3% (26563/31158) | 88.8% (27215/30654) |
| Failures | 3528 | 2459 |
| Errors | 160 | 157 |
| Net improvement | — | **+652 tests** |

Note: test totals vary between runs due to BrokerPool shutdown timeouts.

## Test plan

- [x] XQTS HEAD full suite run shows improvement from 85.3% to 88.8%
- [x] exist-core unit tests: BUILD SUCCESS (0 failures)
- [x] Codacy/PMD local analysis: no new warnings introduced
- [ ] CI verification

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java [hunk 1]
```diff
@@ -195,21 +195,27 @@ public AtomicValue convertTo(final int requiredType) throws XPathException {
 
     public DecimalValue toDecimalValue() throws XPathException {
         if (isNaN() || isInfinite()) {
-            throw conversionError(Type.DECIMAL);
```
Comment (reinhapa) [inline exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java:198]: Why no longer using the existing `conversionError()`method?

File: exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java [hunk 2]
```diff
@@ -195,21 +195,27 @@ public AtomicValue convertTo(final int requiredType) throws XPathException {
...
+            throw new XPathException(getExpression(), ErrorCodes.FOCA0002,
+                    "Cannot convert " + Type.getTypeName(getType()) + "('" + getStringValue()
+                    + "') to " + Type.getTypeName(Type.INTEGER));
         }
         return new IntegerValue(getExpression(), (long) value);
     }
 
     public IntegerValue toIntegerSubType(final int subType) throws XPathException {
         if (isNaN() || isInfinite()) {
-            throw conversionError(subType);
```
Comment (reinhapa) [inline exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java:212]: Why no longer using the existing `conversionError()`method?

File: exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java [hunk 3]
```diff
@@ -195,21 +195,27 @@ public AtomicValue convertTo(final int requiredType) throws XPathException {
...
+            throw new XPathException(getExpression(), ErrorCodes.FOCA0002,
+                    "Cannot convert " + Type.getTypeName(getType()) + "('" + getStringValue()
+                    + "') to " + Type.getTypeName(Type.DECIMAL));
         }
         return new DecimalValue(getExpression(), BigDecimal.valueOf(value));
     }
 
     public IntegerValue toIntegerValue() throws XPathException {
         if (isNaN() || isInfinite()) {
-            throw conversionError(Type.INTEGER);
```
Comment (reinhapa) [inline exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java:205]: Why no longer using the existing `conversionError()`method?

File: exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java (hunk 1 above)
Comment (reinhapa) [inline exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java:198]: why was the method call `conversionError()` removed and code duplication introduced? we could pass in an additional error code for the other usages instead in order to reduce code duplication.

File: exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java (hunk 3 above)
Comment (reinhapa) [inline exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java:205]: same as in `toDecimalValue()`

File: exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java (hunk 2 above)
Comment (reinhapa) [inline exist-core/src/main/java/org/exist/xquery/value/DoubleValue.java:212]: same as in `toDecimalValue()`

File: exist-core/src/main/java/org/exist/xquery/value/Type.java [hunk 4]
```diff
@@ -716,6 +716,117 @@ public static int primitiveTypeOf(final int type) throws IllegalArgumentExceptio
...
+     *
+     * <p>If the cast is impossible, the caller should raise XPTY0004 rather than
+     * attempting the cast (which would incorrectly raise FORG0001).</p>
+     *
+     * @param sourceType the type constant of the source type
+     * @param targetType the type constant of the target type
+     *
+     * @return true if the cast may succeed (for some values), false if the cast can never succeed
+     */
+    public static boolean isCastable(final int sourceType, final int targetType) {
```
Comment (reinhapa) [inline exist-core/src/main/java/org/exist/xquery/value/Type.java:733]: Please address the the  NPath complexity of 8960, current threshold is 200.

**Review by reinhapa — CHANGES_REQUESTED** [review]

**Comment by joewiz** [comment]
[This response was co-authored with Claude Code. -Joe]

Thanks for the review, @reinhapa! Addressed both points in f92f3e6:

1. **DoubleValue**: Added an overloaded `conversionError(ErrorCode, int)` so the NaN/INF paths reuse the existing helper instead of inlining. The original single-arg `conversionError(int)` now delegates to it.

2. **Type.isCastable()**: Extracted the primitive casting table into `isPrimitiveCastable()` and factored the repeated type-group checks into `isNumericTarget()`, `isDateTimeTarget()`, `isGregorianTarget()`, and `isStringOrUntypedAtomic()` helpers. NPath complexity drops from 8960 to below the 200 threshold (confirmed via local Codacy/PMD run).

**Review by reinhapa — APPROVED** [review]

**Review by duncdrum — APPROVED** [review]

**Comment by joewiz** [comment]
[This response was co-authored with Claude Code. -Joe]

The `CDataIntergationTest.cdataWebDavApi` failure is unrelated to this PR's changes. The test passes locally on both `develop` and this branch. This appears to be a flaky CI test (likely timing/port-binding in the embedded WebDAV server).

**Comment by joewiz** [comment]
[This comment was co-authored with Claude Code. -Joe]

Closing — superseded by #6207 (v2/xq31-compliance-fixes).

This work has been consolidated into a clean v2/ branch as part of the eXist-db 7.0 PR reorganization. The new PR includes all commits from this PR plus additional related work, with reviewer feedback incorporated where applicable. See the [reviewer guide](https://github.com/joewiz/exist/blob/next-v2/taskings/v2-reviewer-guide.md) for the full context.
